tsl/src/compression/api.c

```diff
@@ -1075,128 +1061,692 @@ tsl_get_compressed_chunk_index_for_recompression(PG_FUNCTION_ARGS)
+			for (int i = 0; i < nsegmentby_cols; i++)
+				current_segment[i] = palloc(sizeof(CompressedSegmentInfo));
+			initialize_segment(&decompressor,
+							   current_segment,
+							   uncompressed_chunk,
+							   uncompressed_slot);
+		}
+		else
+		{
+			/* check if there is change in segmentby value */
+			is_new_segment = decompress_segment_changed_group(current_segment,
+															  uncompressed_slot,
+															  nsegmentby_cols);
+		}
+		if (is_new_segment)
+		{
+			if (compressed_tuple_found)
+			{
+				/* there exists a previously fetched compressed row */
+				Datum meta_count = 0;
+				Datum seq_num = 0;
+				fetch_segment_metadata(&decompressor,
+									   *prev_compressed_tuple,
+									   &meta_count,
+									   &seq_num);
+				/*
+				 * Check if this compressed row has mx rows per compression.
+				 * If total rows per compression exceeds MAX_ROWS_PER_COMPRESSION
+				 * we either decompress all compressed rows within same segment
+				 * OR ignore MAX_ROWS_PER_COMPRESSION and append more rows to
+				 * existing compressed row.
+				 */
+				if (((meta_count + total_uncompressed_rows_per_segment) >=
+					 MAX_ROWS_PER_COMPRESSION) &&
+					ts_guc_ignore_max_rows_per_compression == false)
+				{
+					int segments_decompressed =
+						fetch_compressed_tuples_per_segment(&decompressor,
+															uncompressed_chunk,
+															current_segment,
+															seq_num,
+															nsegmentby_cols,
+															tuplestorecxt->tuplestore);
+					total_segments_decompressed += segments_decompressed;
+				}
+				/* perform decompression and save rows into tuplestore */
+				row_compressor.update_sequence = false;
+				row_compressor.sequence_num = decompress_in_tuplestore(&decompressor,
+																	   *prev_compressed_tuple,
+																	   tuplestorecxt->tuplestore);
+				total_segments_decompressed++;
+				compressed_tuple_found = false;
+			}
+			/* perform compression on accumulated rows present in tuplestore */
+			tuplesort_performsort(tuplestorecxt->tuplestore);
+			recompress_segment(tuplestorecxt->tuplestore, decompressor.out_rel, &row_compressor);
+			cleanup_tuplestorecxt(tuplestorecxt);
+			tuplestorecxt = initialize_tuplestore(decompressor.out_rel, keys, n_keys);
+			total_uncompressed_rows_per_segment = 0;
+			if (prev_compressed_tuple)
 			{
-				valid = false;
-				break;
+				pfree(prev_compressed_tuple);
+				prev_compressed_tuple = NULL;
 			}
+			/* update new segment */
+			decompress_segment_update_current_segment(current_segment,
+													  uncompressed_slot,
+													  nsegmentby_cols);
+			last_segment_tuple_found = false;
 		}
-		if (valid)
+		/* build scan keys to do lookup into compressed chunks */
+		seg_order_by_scankeys = build_segment_order_by_scankeys(&decompressor,
+																uncompressed_chunk,
+																uncompressed_slot,
+																&num_scankeys);
+
+		TableScanDesc heapScan = table_beginscan(decompressor.in_rel,
```

Comment:
(same thing here, maybe doing an indexscan instead of heapscan on the compressed chunk)

Reply:
At this place, its not possible to use index scan. I need to build scankeys on segmentby and orderby columns. There is no index defined on compressed chunk which has order by columns.
So keeping the code as is.